java/ql/lib/semmle/code/java/dataflow/ListOfConstantsSanitizer.qll

```diff
@@ -0,0 +1,357 @@
+  // f("a", "b", "c")
+  forex(Expr e | e = mc.getAnArgument() |
+    not e.getType() instanceof Array and e.isCompileTimeConstant()
+  )
+  or
+  // String[] a = {"a", "b", "c"};
+  // f(a)
+  mc.getNumArgument() = 1 and
+  mc.getArgument(0).getType() instanceof Array and
+  exists(ArrayInit arr | DataFlow::localExprFlow(arr, mc.getArgument(0)) |
+    forex(Expr e | e = arr.getAnInit() | e.isCompileTimeConstant())
+  )
+}
+
+/** The name of a class implementing `java.util.Collection`. */
+class CollectionClass extends string {
+  CollectionClass() { this = ["List", "Set"] }
+}
+
+/** Classes for `java.util.List` and `java.util.Set`. */
+module Collection {
+  private class CollectionContainsCall extends MethodCall {
+    CollectionClass collectionClass;
+
+    /** Gets whether the collection is a "List" or a "Set". */
+    CollectionClass getCollectionClass() { result = collectionClass }
+
+    CollectionContainsCall() {
+      exists(Method m |
+        this.getMethod() = m and
+        m.hasName("contains") and
+        m.getDeclaringType()
+            .getSourceDeclaration()
+            .getASourceSupertype*()
+            .hasQualifiedName("java.util", collectionClass)
+      )
+    }
+  }
+
+  /**
+   * A call with a collection of class `collectionClass` as argument `arg` (or
+   * as qualifier, if `arg` is -1) which will not add any new non-constant
+   * elements to it.
+   */
+  abstract class SafeCall extends Call {
+    int arg;
+    CollectionClass collectionClass;
+
+    SafeCall() {
+      arg = -1 and exists(this.getQualifier())
+      or
+      exists(this.getArgument(arg))
+    }
+
+    /** Gets whether the collection is a "List" or a "Set". */
+    CollectionClass getCollectionClass() { result = collectionClass }
+
+    /** Gets the argument index, or -1 for the qualifier. */
+    int getArg() { result = arg }
+  }
+
+  private class AddConstantElement extends SafeCall, MethodCall {
+    AddConstantElement() {
+      arg = -1 and
+      exists(Method m, RefType t |
+        this.getMethod() = m and
+        t = m.getDeclaringType().getSourceDeclaration().getASourceSupertype*()
+      |
+        collectionClass = "List" and
+        t.hasQualifiedName("java.util", "List") and
+        m.getName() = ["add", "addFirst", "addLast"] and
+        this.getArgument(m.getNumberOfParameters() - 1).isCompileTimeConstant()
+        or
+        collectionClass = "Set" and
+        t.hasQualifiedName("java.util", "Set") and
+        m.getName() = "add" and
+        this.getArgument(0).isCompileTimeConstant()
+      )
+    }
+  }
+
+  private class UnmodifiableCollection extends SafeCall, MethodCall {
+    UnmodifiableCollection() {
+      arg = 0 and
+      exists(Method m |
+        this.getMethod() = m and
+        m.hasName("unmodifiable" + collectionClass) and
+        m.getDeclaringType()
+            .getSourceDeclaration()
+            .getASourceSupertype*()
+            .hasQualifiedName("java.util", "Collections")
+      )
```

Comment:
These are static methods, so this can be simplified:
```suggestion
      this.getMethod().hasQualifiedName("java.util", "Collections", "unmodifiable" + collectionClass)
```